### PR TITLE
Add scsi_tape parser

### PR DESCRIPTION
**Pull request by luizluca**
For prometheus/node_exporter#2044

Signed-off-by: Luiz Angelo Daros de Luca <luizluca@gmail.com>

### CODE REVIEW

File: fixtures.ttar [hunk 1]
```diff
@@ -6551,3 +7047,7 @@ Lines: 1
 4733
 Mode: 644
 # ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+Path: fixtures/.unpacked
+Lines: 0
+Mode: 644
+# ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
```
Comment (SuperQ) [inline fixtures.ttar:7053]: This shouldn't be in here. There's a `make update_fixtures` helper that takes care of re-packing the fixtures.

File: fixtures.ttar [hunk 2]
```diff
@@ -5254,6 +5721,35 @@ Mode: 644
 Directory: fixtures/sys/devices/system/cpu/cpufreq/policy1
 Mode: 755
 # ttar - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - - -
+Directory: fixtures/sys/devices/system/node
```
Comment (SuperQ) [inline fixtures.ttar:5724]: Looks like some re-ordering is happening here.

Comment (SuperQ) [reply fixtures.ttar:5724]: I double checked, and yes, this is correct ordering.

I guess we should add a repack fixtures test to the CI pipeline.

Comment (SuperQ) [reply fixtures.ttar:5724]: https://github.com/prometheus/procfs/pull/383 should fix this up.

File: sysfs/class_scsitape.go [hunk 3]
```diff
@@ -0,0 +1,140 @@
+// Copyright 2020 The Prometheus Authors
```
Comment (SuperQ) [inline sysfs/class_scsitape.go:1]:
```suggestion
// Copyright 2021 The Prometheus Authors
```

File: sysfs/class_scsitape_test.go [hunk 4]
```diff
@@ -0,0 +1,56 @@
+// Copyright 2019 The Prometheus Authors
```
Comment (SuperQ) [inline sysfs/class_scsitape_test.go:1]:
```suggestion
// Copyright 2021 The Prometheus Authors
```

File: sysfs/class_scsitape.go (hunk 3 above)
Comment (luizluca) [reply sysfs/class_scsitape.go]: Done

File: sysfs/class_scsitape_test.go (hunk 4 above)
Comment (luizluca) [reply sysfs/class_scsitape_test.go]: Done

File: fixtures.ttar (hunk 2 above)
Comment (luizluca) [reply fixtures.ttar:5724]: I'm not sure what of if I need to fix something here. I forced-pushed after make update_fixtures

File: fixtures.ttar (hunk 1 above)
Comment (luizluca) [reply fixtures.ttar]: make update_fixtures done

**Comment by luizluca** [comment]
Is there anything else pending?

File: sysfs/class_scsitape.go [hunk 5]
```diff
@@ -0,0 +1,140 @@
...
+// ScsiTapeClass parses st[0-9]+ devices in /sys/class/scsi_tape.
+func (fs FS) ScsiTapeClass() (ScsiTapeClass, error) {
+	path := fs.sys.Path(scsitapeClassPath)
+
+	dirs, err := ioutil.ReadDir(path)
+	if err != nil {
+		return nil, err
+	}
+
+	// TODO: unpredictable
```
Comment (discordianfish) [inline sysfs/class_scsitape.go:58]: What does this comment mean?

File: sysfs/class_scsitape.go [hunk 6]
```diff
@@ -0,0 +1,140 @@
...
+import (
+	"fmt"
+	"io/ioutil"
+	"path/filepath"
+	"regexp"
+
+	"github.com/prometheus/procfs/internal/util"
+)
+
+const scsitapeClassPath = "class/scsi_tape"
```
Comment (discordianfish) [inline sysfs/class_scsitape.go:27]: Should be scsiTapeClassPath

File: sysfs/class_scsitape.go [hunk 7]
```diff
@@ -0,0 +1,140 @@
...
+
+type ScsiTapeCounters struct {
```
Comment (discordianfish) [inline sysfs/class_scsitape.go:29]: Should be SCSITapeCounters. Acronyms should be all uppercase as of golang best practices

File: sysfs/class_scsitape.go [hunk 8]
```diff
@@ -0,0 +1,140 @@
...
+
+type ScsiTape struct {
+	Name     string           // /sys/class/scsi_tape/<Name>
+	Counters ScsiTapeCounters // /sys/class/scsi_tape/<Name>/statistics/*
+}
+
+type ScsiTapeClass map[string]ScsiTape
+
+// ScsiTapeClass parses st[0-9]+ devices in /sys/class/scsi_tape.
+func (fs FS) ScsiTapeClass() (ScsiTapeClass, error) {
```
Comment (discordianfish) [inline sysfs/class_scsitape.go:50]: Same in functions etc

File: sysfs/class_scsitape.go (hunk 5 above)
Comment (luizluca) [reply sysfs/class_scsitape.go:58]: I replaced it with:

```
// There are n?st[0-9]+[a-b]? variants depending on device features.
// n/2 is probably overestimated but never underestimated
```

I'm new to golang but latter I found out that the "make" here might not matter much for a map. It's just for internal preallocatted structures. I'll divide by 2 for a better guess.

File: sysfs/class_scsitape.go (hunk 6 above)
Comment (luizluca) [reply sysfs/class_scsitape.go:27]: ok

File: sysfs/class_scsitape.go (hunk 7 above)
Comment (luizluca) [reply sysfs/class_scsitape.go:29]: ok

File: sysfs/class_scsitape.go (hunk 8 above)
Comment (luizluca) [reply sysfs/class_scsitape.go:50]: ok

File: sysfs/class_scsitape.go [hunk 9]
```diff
@@ -0,0 +1,140 @@
...
+
+	for _, f := range files {
+		name := filepath.Join(path, f.Name())
+		value, err := util.SysReadFile(name)
+		if err != nil {
+			return nil, fmt.Errorf("failed to read file %q: %w", name, err)
+		}
+
+		vp := util.NewValueParser(value)
+		// '<,'>s/^\(\(.\)\([^_]*\)\(_\(.\)\([^_]*\)\)\(_\(.\)\([^_]*\)\)*\)/^I^Icase "\1":\r^I^I^Icounters.\U\2\e\3\U\5\e\6\U\8\e\9 = *vp.PUInt64()/
```
[the remark this answers is not in the document]
Comment (luizluca) [reply sysfs/class_scsitape.go:109]: It's a vim command to convert the statistics filenames in sysfs into the code you see bellow. I'll remove it.

Something like:

```
# from inside the stats dir:
:r! ls
# select the names in visual and run:
'<,'>s/^\(\(.\)\([^_]*\)\(_\(.\)\([^_]*\)\)\(_\(.\)\([^_]*\)\)*\)/^I^Icase "\1":\r^I^I^Icounters.\U\2\e\3\U\5\e\6\U\8\e\9 = *vp.PUInt64()/
```

**Comment by luizluca** [comment]
I just forced-push the review suggestions. I'll do the same for node_exporter collector.

**Review by SuperQ — APPROVED** [review]
LGTM

**Review by discordianfish — APPROVED** [review]

**Comment by discordianfish** [comment]
Awesome, thanks @luizluca!

**Pull request merged**